src/dms/backend/EdFi.DataManagementService.Backend.Postgresql/Deploy/Scripts/0010_Create_Insert_References_Procedure.sql

```diff
@@ -3,64 +3,197 @@
 -- The Ed-Fi Alliance licenses this file to you under the Apache License, Version 2.0.
 -- See the LICENSE and NOTICES files in the project root for more information.
 
--- Attempts to insert references into Reference table
--- If any referentialId is invalid, rolls back and returns an array of invalid referentialIds
+-- Upserts document references for a parent document.
+-- The provided referentialIds MUST be unique
+-- Returns a success flag plus the invalid referentialIds if unsuccessful.
 CREATE OR REPLACE FUNCTION dms.InsertReferences(
-    parentDocumentIds BIGINT[],
-    parentDocumentPartitionKeys SMALLINT[],
-    referentialIds UUID[],
-    referentialPartitionKeys SMALLINT[]
-) RETURNS TABLE (ReferentialId UUID)
-LANGUAGE plpgsql AS
+    p_parentDocumentId BIGINT,
+    p_parentDocumentPartitionKey SMALLINT,
+    p_referentialIds UUID[],
+    p_referentialPartitionKeys SMALLINT[],
+    p_isPureInsert BOOLEAN DEFAULT FALSE
+) RETURNS TABLE (
+    success BOOLEAN,
+    invalid_ids UUID[]
+)
+LANGUAGE plpgsql
+AS
 $$
 DECLARE
-    constraintErrorName text;
+    reference_partition TEXT;
+    current_session_id INTEGER := pg_backend_pid();
+    needs_upsert BOOLEAN := TRUE;
 BEGIN
+    -- Reuse the unlogged staging table across calls, discard any leftovers from prior aborted executions.
+    DELETE FROM dms.ReferenceStage
+    WHERE SessionId = current_session_id;
 
-    -- First clear out all the existing references, as they may have changed
-    DELETE from dms.Reference r
-    USING unnest(parentDocumentIds, parentDocumentPartitionKeys) as d (Id, DocumentPartitionKey)
-    WHERE d.Id = r.ParentDocumentId AND d.DocumentPartitionKey = r.ParentDocumentPartitionKey;
-
-    INSERT INTO dms.Reference (
-        ParentDocumentId,
-        ParentDocumentPartitionKey,
-        ReferentialId,
-        ReferentialPartitionKey,
-        ReferencedDocumentId,
-        ReferencedDocumentPartitionKey
+    WITH staged AS (
```

Comment:
Stages references into a permanent unlogged table. Unlogged = no WAL, and the data lifetime is in the transaction.